cuda/create_grid.cu

```diff
@@ -133,29 +141,128 @@ Grid *quadtree_grid(Point *points, int count, pair<float, float> bottom_left_cor
 	pair<float, float> upperBound = make_pair(x2, y2);
 	pair<float, float> lowerBound = make_pair(x1, y1);
 
-	return new Grid(bl_grid, br_grid, tl_grid, tr_grid, points, upperBound, lowerBound, count);
+	Grid *root_grid = new Grid(bl_grid, br_grid, tl_grid, tr_grid, points, upperBound, lowerBound, count, parent, id);
+	grid_map[id] = root_grid;
+
+	if (bl_grid)
+		bl_grid->parent = root_grid;
+	if (br_grid)
+		br_grid->parent = root_grid;
+	if (tl_grid)
+		tl_grid->parent = root_grid;
+	if (tr_grid)
+		tr_grid->parent = root_grid;
+
+	return root_grid;
 }
 
-int main(int argc, char *argv[]) {
-	if (argc < 3) {
+// Function returns the quadrant id of the given point
+int search_quadrant(Point target_point, const vector<QuadrantBoundary> &boundaries)
+{
+	QuadrantBoundary *d_boundaries;
+	cudaMalloc(&d_boundaries, boundaries.size() * sizeof(QuadrantBoundary));
+	cudaMemcpy(d_boundaries, boundaries.data(), boundaries.size() * sizeof(QuadrantBoundary), cudaMemcpyHostToDevice);
+
+	Point *d_target_point;
+	cudaMalloc(&d_target_point, sizeof(Point));
+	cudaMemcpy(d_target_point, &target_point, sizeof(Point), cudaMemcpyHostToDevice);
+
+	int *d_result;
+	cudaMalloc(&d_result, sizeof(int));
+
+	int init_value = -1;
+	cudaMemcpy(d_result, &init_value, sizeof(int), cudaMemcpyHostToDevice);
+
+	int block_size = 256;
+	int num_blocks = 16;
+	quadrant_search<<<num_blocks, block_size>>>(d_target_point, d_boundaries, boundaries.size(), d_result);
+
+	int result;
+	cudaMemcpy(&result, d_result, sizeof(int), cudaMemcpyDeviceToHost);
+
+	cudaFree(d_boundaries);
+	cudaFree(d_target_point);
+	cudaFree(d_result);
+
+	return (result == -1) ? -1 : result; // Return -1 if point not found in any quadrant
+}
+
+void insert_point(Point new_point, Grid *root_grid, vector<QuadrantBoundary> &boundaries, unordered_map<int, Grid *> &grid_map)
+{
+	// First, locate the quadrant ID where the new point should go
+	int quadrant_id = search_quadrant(new_point, boundaries);
+	if (quadrant_id == -1)
+	{
+		printf("The point is outside the grid boundaries and cannot be inserted.\n");
+		return;
+	}
+
+	// Access the target grid from the unordered map
+	Grid *target_grid = grid_map[quadrant_id];
+
+	// Check if the point already exists in the quadrant
+	for (int i = 0; i < target_grid->count; i++)
+	{
+		if (target_grid->points[i].x == new_point.x && target_grid->points[i].y == new_point.y)
+		{
+			printf("The point already exists in the grid.\n");
+			return;
+		}
+	}
+
+	// Add the point to the grid
+	Point *new_points = (Point *)malloc((target_grid->count + 1) * sizeof(Point));
+	memcpy(new_points, target_grid->points, target_grid->count * sizeof(Point));
+	new_points[target_grid->count] = new_point;
+	free(target_grid->points);
+	target_grid->points = new_points;
+	target_grid->count++;
+
+	// Propagate count increment to all parent nodes
+	Grid *parent_grid = target_grid->parent;
+	while (parent_grid)
+	{
+		parent_grid->count++;
+		parent_grid = parent_grid->parent;
```

Comment:
Also need to insert the point into the parent_grid->points array. Has to be added at every level.